### PR TITLE
Some changes to HFile.select() and SingleDetTriggers

### PR DESCRIPTION
**Pull request by GarethCabournDavies**
This is an attempt to sort #4524

These changes allow:
 - any of the rankings from ranking.py to be used in the `HFile.select()` function
 - A pre-sort directly within SingleDetectorTriggers, so that the `HFile.select()` functionality can used for simple things easily
 - A pre-cut in `mask_to_n_loudest_clustered_events` in cases where this is appropriate

So far as I can test, this has not broken previous functionality except for one usage, which is the filter string option in `pycbc_plot_singles_vs_params`

### CODE REVIEW

File: pycbc/io/hdf.py
Comment (ahnitz) [inline pycbc/io/hdf.py:128]: @GarethCabournDavies This seems like a odd shoehorn into this class to me. The idea is that this class should be general and only really build upon h5py / numpy and so can be used for general purposes where more careful access of file informatoin is needed.  Is there a way to avoid explicitly calling a module from ranking.py? Perhaps by extending the options to this function?

**Review by ahnitz — CHANGES_REQUESTED** [review]

File: pycbc/io/hdf.py
Comment (ahnitz) [reply pycbc/io/hdf.py:128]: @GarethCabournDavies Ok, I think one approach would be to add something like a 'derived' keyword where on can provide a dictionary -> function mapping (could b lambda function to allow more complicated function interface like you might have with the ranking one. 

So something like the following input. 

derived = {'derived_name': function} 

And in this case, the 'required' keys should just be expanded with the ones you know your functions need. I see you added this data to ranking.py, so I think where you are calling the select method from should have this information.

Comment (GarethCabournDavies) [reply pycbc/io/hdf.py:128]: That makes sense - I think this will still require the dictionary setup I used for passing to the derived (ranking) functions, as otherwise there could be duplicated datasets which we should avoid, but this does add a layer of generality

Comment (GarethCabournDavies) [reply pycbc/io/hdf.py]: The select() function has been updated to be agnistic of any other module.

SingleDetTriggers still calls through for defined ranking statistics - as SingleDetTriggers already has this happening for certain attributes, I thought this should be okay still

**Comment by GarethCabournDavies** [comment]
Testing on the updates to the executables which use these are ongoing. I will point to the success (or make changes) when they are complete

File: pycbc/io/hdf.py
Comment (ahnitz) [reply pycbc/io/hdf.py]: SingleDetTriggers is specialized so I think it makes sense there.

**Comment by GarethCabournDavies** [comment]
I am going to try and un-break the filter string option as the next step in this

File: bin/all_sky_search/pycbc_bin_trigger_rates_dq
Comment (GarethCabournDavies) [inline bin/all_sky_search/pycbc_bin_trigger_rates_dq:57]: Note that this example is pretty much the reasoning behind this change - to clean up codes like this

File: bin/minifollowups/pycbc_injection_minifollowup
Comment (GarethCabournDavies) [inline bin/minifollowups/pycbc_injection_minifollowup:309]: The changes here were possible before, but this is just to implement the select() function and make things quite a bit faster

File: bin/plotting/pycbc_plot_singles_vs_params
Comment (GarethCabournDavies) [inline bin/plotting/pycbc_plot_singles_vs_params:164]: This can be removed if wanted - it is basically a safety for if there aren't many triggers - this is a consequence of applying the `--min-z`

File: examples/gw150914/PyCBCInspiral.ipynb
Comment (GarethCabournDavies) [inline examples/gw150914/PyCBCInspiral.ipynb:919]: I'm not actually sure this will work to directly change this like this

**Comment by GarethCabournDavies** [comment]
Okay, I think this is ready to review now

File: pycbc/io/hdf.py
Comment (spxiwh) [inline pycbc/io/hdf.py:676]: Could/should these lines be computed using `trigs_f.select` as well? Would be good to avoid reading in all trigger's information if we don't need to in this method.

**Review by spxiwh — COMMENTED** [review]
Thanks Gareth. There's some comments/suggestions from me here, but I would like to have @ahnitz 's input before merging this. In general no big worries here, just some aesthetics, and a couple of ideas to help future proof.

File: bin/plotting/pycbc_plot_singles_vs_params
Comment (GarethCabournDavies) [reply bin/plotting/pycbc_plot_singles_vs_params:87]: I left this in as I could see the situation where applying a minimum SNR would be different to applying a minimum z_var value (e.g. when psdvar is in play) so thought it best to leave both in.

I can simply get rid of min_snr if that is wanted

File: pycbc/io/hdf.py
Comment (GarethCabournDavies) [reply pycbc/io/hdf.py:95]: It doesn't, this itself was a way to avoid a problem which came up if using `self[group][ds]`

File: pycbc/io/hdf.py
Comment (GarethCabournDavies) [reply pycbc/io/hdf.py:48]: I have specified the arguments

File: pycbc/io/hdf.py
Comment (GarethCabournDavies) [reply pycbc/io/hdf.py:162]: Okay, so in order to get this to consistently return the same number of outputs, I think that there will be two outputs: the indices, and the tuple of datasets, which could be empty if return_data is False

So I will remove the return_indices option

File: pycbc/io/hdf.py
Comment (GarethCabournDavies) [reply pycbc/io/hdf.py:122]: See the below comment

File: bin/plotting/pycbc_plot_singles_vs_params
Comment (GarethCabournDavies) [reply bin/plotting/pycbc_plot_singles_vs_params:143]: Fixed, (as is line below)

File: pycbc/io/hdf.py
Comment (GarethCabournDavies) [reply pycbc/io/hdf.py:532]: That would lose some functionality compared to what is in there at the moment. As it would then restrict us to the ranking functions only. We could use a tuple as the value, which defines the function and the datasets required, so we are not hard-coded only to those

e.g.
```
indices, tuple_out = trig_f.select(
    a_function_which_uses_end_time_and_newsnr_sgveto,
    'end_time',
    'newsnr_sgveto'
    group=opts.detector
    derived={'newsnr_sgveto': (ranking.sngls_ranking_function_dict['newsnr_sgveto'], 
                               ranking.required_datasets['newsnr_sgveto'])
)
```
which would return a tuple with 'end_time' and 'newsnr_sgveto'

File: pycbc/io/hdf.py
Comment (GarethCabournDavies) [reply pycbc/io/hdf.py:543]: `pycbc_plot_singles_vs_params`, `pycbc_process_sngls` and `pycbc_fit_sngl_trigs` all use the `filter_func` option (from git grepping), I don't know whether all these are still needing to use it though

I think this could be sent to the select() function, but wanted to sort other parts first

Comment (GarethCabournDavies) [reply pycbc/io/hdf.py:543]: One complication here is where the filter_func uses derived datasets, or template bank parameters

File: examples/gw150914/PyCBCInspiral.ipynb
Comment (GarethCabournDavies) [reply examples/gw150914/PyCBCInspiral.ipynb:919]: I had a look and it was wrong - however the notebook instructions are to use a specific version of pycbc (v1.7.5), and the notebook doesn't work with the current master for other reasons.

So I think it may be best for me to completely remove any changes to this file and not worry about it

File: pycbc/events/ranking.py
Comment (ahnitz) [inline pycbc/events/ranking.py:303]: Does it make this linked to the function themselves? I'm just wondering if this is really maintainable.

**Comment by ahnitz** [comment]
@GarethCabournDavies I don't see too much to comment on anymore,, except the one thing. Also can you comment or resolve the issues above? It's not quite clear what is still be working on and what the choice made was in response.

**Comment by GarethCabournDavies** [comment]
> @GarethCabournDavies I don't see too much to comment on anymore,, except the one thing. Also can you comment or resolve the issues above? It's not quite clear what is still be working on and what the choice made was in response.

I was leaving them open until I have finished testing and committed, but will resolve as soon as possible

File: pycbc/events/ranking.py
Comment (GarethCabournDavies) [reply pycbc/events/ranking.py:303]: These are not linked, the dictionary will need updating as more functions are added into this module.

One way to add this would be if we define the dictionary before the functions, and all functions have an assert that the required_datasets are in trigs. Then if another function is made (likely copied and adapted), then people will naturally add the list into this dictionary. This doesn't make it more maintainable, but does add a psychological incentive to add the list

Comment (GarethCabournDavies) [reply pycbc/events/ranking.py:303]: I've added the check, so anyone using `get_sngls_ranking_from_trigs` without adding the required datasets will have a problem

File: pycbc/io/hdf.py
Comment (GarethCabournDavies) [reply pycbc/io/hdf.py]: I have made this change

File: bin/plotting/pycbc_plot_singles_vs_params
Comment (GarethCabournDavies) [reply bin/plotting/pycbc_plot_singles_vs_params]: removed

**Comment by GarethCabournDavies** [comment]
Noticed an error in the way pre-masks are handled alongside the new filter_rank in SingleDetTriggers (it's not actually used anywhere in the codebase, but just to be safe), I think I have the fix ready, but this is not to merge before then

File: pycbc/events/ranking.py
Comment (GarethCabournDavies) [reply pycbc/events/ranking.py:303]: I've had to abandon this idea, as the `get_sngls_ranking_from_trigs` function is used by quite a few different input classes for `trigs`, so making the check general enough was over-complicated (in my mind at least)

File: pycbc/io/hdf.py
Comment (GarethCabournDavies) [reply pycbc/io/hdf.py:556]: I'll look at this after this has been merged - there has already been quite a bit of scope creep of this PR

**Comment by GarethCabournDavies** [comment]
Fix to the above is in, so please approve if you are happy

**Comment by tdent** [comment]
Where are we with this, looks like just waiting for review?

**Comment by GarethCabournDavies** [comment]
> Where are we with this, looks like just waiting for review?

Yup - the last update was in the day or so before christmas break, so not many working days to have had a look at this

**Comment by GarethCabournDavies** [comment]
Poke for review on this

**Review by spxiwh — APPROVED** [review]
I'm happy with the changes now. I think this would also require sign-off from @ahnitz  before it could be merged.

**Comment by spxiwh** [comment]
... And there's now conflicts.

**Review by ahnitz — APPROVED** [review]
Sorry for the late reply, yes, I think this has addressed the issues I originally raised, and I don't see anything too major. Always cleanup can be done, but I think this leaves it in a better state than it was.

**Pull request merged**
